### PR TITLE
Improve TeamTalk support

### PR DESCRIPTION
**Pull request by codeofdusk**
### Link to issue number:
Related to #9548 (TeamTalk 3.x support removed).

### Summary of the issue:
NVDA's TeamTalk support is minimal and outdated.

### Description of how this pull request fixes the issue:
This PR introduces a new TeamTalk app module, based on an add-on by @dglee42 and @tspivey. Changes made since the original add-on:

* Python 3 compatible.
* Added translator comments.
* Various code style fixes, replaced magic numbers with named constants for easier maintenance and readability.
* The VU progress bar is also disabled on newer client versions.

This PR drops support for TeamTalk 3.x, which is at least ten years old and almost definitely has extremely low usage.

### Testing performed:
Tested the new app module on TeamTalk 5.4. Verified that buttons and controls are properly labeled, the read status script is functional, and no spurious progress updates are presented when the VU meter is enabled.

### Known issues with pull request:
None.

### Change log entry:
== New Features ==
- NVDA's TeamTalk support works with newer versions of the client. (#10071)

In the "AppModules removed" section of Changes for Developers, add:

- TeamTalk 3.x.

### CODE REVIEW

**Comment by codeofdusk** [comment]
Cc @feerrenrut @josephsl @leonardder

**Review by josephsl — CHANGES_REQUESTED** [review]
Hi,

Few general things:

1. Can you merge in latest master branch? Although you don't have to, merging it makes reviewing easier. I ask because of the fact that master branch contains vision enhancer framework.
2. For TeamTalk 4 app module, can you still add a copyright header for consistency? At least we can figure out who did what based on git log, someone may wonder who did what without knowing about Git.
3. User guide: is the toolbar announcement command specific to NVDA or TeamTalk? If the former, I advise adding a section in the user guide about it.

Thanks.

**Comment by codeofdusk** [comment]
@josephsl Requested changes made, and a user guide section was already added.

**Review by josephsl — APPROVED** [review]

**Comment by josephsl** [comment]
Hi,

Even though I approved it, I think it would be best to hear from @DGLee42 and @TSpivey for their opinions, since they are the ones who provided foundation for this PR.

Thanks

**Review by LeonarddeR — CHANGES_REQUESTED** [review]

File: source/appModules/teamtalk5classic.py [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+LBL_CONNECTED = _("Connected")
+# Translators: Status text in the TeamTalk voice conferencing client.
+LBL_PTT = _("Push To Talk")
+# Translators: Status text in the TeamTalk voice conferencing client.
+LBL_VOX = _("Voice Activation")
+# Translators: Status text in the TeamTalk voice conferencing client.
+LBL_VIDEO = _("Video")
+# Translators: Status text in the TeamTalk voice conferencing client.
+# Note: this string is slightly different from the actual toolbar text
+# to avoid a translatable string conflict.
```
[the remark this answers is not in the document]
Comment (codeofdusk) [reply source/appModules/teamtalk5classic.py]: `scons checkPot` complained that a translator comment existed for the string "Desktop" even though it was ignored (I think conflicted with another string elsewhere in NVDA).

[the remark this answers is not in the document]
Comment (leonardder) [reply source/appModules/teamtalk5classic.py]: Does checkpot also fail when using pgettext?

**Comment by codeofdusk** [comment]
Cc @michaelDCurran.

**Comment by dingpengyu** [comment]
I just tested nvda pr10071-18444 in Simplified Chinese and he works fine in Chinese.
thank

**Comment by codeofdusk** [comment]
Any updates on this?

**Comment by michaelDCurran** [comment]
Looks like there is a pending review from @leonardder. If he is happy with the changes and approves then I will review.

File: source/appModules/teamtalk5classic.py [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+	def __init__(self, NVDAObj=None):
+		if not NVDAObj:
+			# Use the IAccessible for the active TeamTalk instance's toolbar.
+			NVDAObj = self.getToolBar()
+			if not NVDAObj:
+				# Translators: Reported when status information is unavailable
+				# in the TeamTalk voice conferencing client.
+				ui.message(_("Status unavailable"))
+				return
+		self.NVDAObj = NVDAObj
```
Comment (leonardder) [inline source/appModules/teamtalk5classic.py]: I'm a bit worried about this being a hard refference, introducing a circular refference below when using `self.info = TTToolBarInfo(self)`

File: source/appModules/teamtalk5classic.py [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+	"""
+
+	def __init__(self, NVDAObj=None):
+		if not NVDAObj:
+			# Use the IAccessible for the active TeamTalk instance's toolbar.
+			NVDAObj = self.getToolBar()
+			if not NVDAObj:
+				# Translators: Reported when status information is unavailable
+				# in the TeamTalk voice conferencing client.
+				ui.message(_("Status unavailable"))
```
Comment (leonardder) [inline source/appModules/teamtalk5classic.py]: When this object is initiated and there is no object, status unavailable is spoken, but creating the object still succeeds. I'd rather have initialisation of the object raise an error, catch that and use the ui.message there instead.

File: source/appModules/teamtalk5classic.py [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+		except IndexError:
+			return ""
+		if name == "sep":
+			return ""
+		return name
+
+
+class TTToolBar(IAccessible):
+	def initOverlayClass(self):
+		self.info = TTToolBarInfo(self)
```
Comment (leonardder) [inline source/appModules/teamtalk5classic.py]: If I understand the code correctly, TTToolBarInfo gets the toolbar itself when it is not initialized with an object. However, on that toolbar object, the overlay will be added and another TTToolBarInfo will be created. I think this code is too complicated. Why not add all the logic from TTToolBarInfo to this overlay instead?

File: source/appModules/teamtalk5classic.py [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+
+
+class TTToolBarItem(IAccessible):
+	def event_valueChange(self):
+		speech.speakObject(self)
```
Comment (leonardder) [inline source/appModules/teamtalk5classic.py]: When does this happen and why has this to be overridden?

File: source/appModules/teamtalk5classic.py [hunk 6]
```diff
@@ -0,0 +1,245 @@
...
+			obj._isWindowUnicode = False
+		# Label channel chat in/out controls.
+		if obj.windowControlID == ID_CHATOUT:
+			# Translators: a label for a read only edit control
+			# in the TeamTalk voice conferencing client where incoming chat
+			# messages are displayed.
+			obj.name = _("Channel messages")
+		elif obj.windowControlID == ID_CHATIN:
+			# Translators: a label for the TeamTalk chat entry field
+			obj.name = _("Type a message to the channel")
```
Comment (leonardder) [inline source/appModules/teamtalk5classic.py]: I think this is pretty verbose

```suggestion
			obj.name = _("Channel input")
```

File: source/appModules/teamtalk5classic.py [hunk 7]
```diff
@@ -0,0 +1,245 @@
...
+		# Better name Add/Remove buttons where they occur.
```
Comment (leonardder) [inline source/appModules/teamtalk5classic.py]: I think > and <, though a bit vague, are suitable names for controls like this. I'd be reluctant to change them.

File: source/appModules/teamtalk5classic.py [hunk 8]
```diff
@@ -0,0 +1,245 @@
...
+		elif obj.windowControlID == ID_TOOLBAR:
+			# Applies to toolbar and its items.
+			if obj.role == controlTypes.ROLE_TOOLBAR:
+				clsList.insert(0, TTToolBar)
+			elif obj.parent and obj.parent.role == controlTypes.ROLE_TOOLBAR:
+				clsList.insert(0, TTToolBarItem)
+
+	@scriptHandler.script(
+		gesture="kb:control+shift+s",
+		description="""Says various on/off states from the toolbar. Depending
```
Comment (leonardder) [inline source/appModules/teamtalk5classic.py]: This is not translatable. I also prefer another style here, that's just my personal preference though.

```suggestion
		description=_(
			"Says various on/off states from the toolbar. Depending "
			"on TeamTalk version, these can include connection status, "
			...
		)
```

File: source/appModules/teamtalk5classic.py [hunk 9]
```diff
@@ -0,0 +1,245 @@
...
+features, whether channel audio is being saved to files,
+and whether the channel message window is showing.
+On one press, just says which are checked (in effect).
+On two presses, says all states.
+		"""
+	)
+	def script_sayToolbarInfo(self, gesture):
+		info = TTToolBarInfo()
+		if not info:
+			return
```
Comment (leonardder) [inline source/appModules/teamtalk5classic.py]: That ui.message really should be here somewhere.

File: source/appModules/teamtalk5classic.py [hunk 10]
```diff
@@ -0,0 +1,245 @@
...
+		scnt = scriptHandler.getLastScriptRepeatCount()
+		if scnt > 0:
+			[
+				speech.speakObject(o)
```
Comment (leonardder) [inline source/appModules/teamtalk5classic.py]: I have the feeling that a list comprehension isn't suitable for this. This will result in a list of Nones. Please change this into a real for loop.

**Review by LeonarddeR — CHANGES_REQUESTED** [review]

**Comment by leonardder** [comment]
@codeofdusk: could you address the review comments?

File: source/appModules/teamtalk5classic.py (hunk 5 above)
Comment (codeofdusk) [reply source/appModules/teamtalk5classic.py]: > When does this happen and why has this to be overridden?

Possibly legacy code for an old client version? This was from the original add-on.

File: source/appModules/teamtalk5classic.py (hunk 3 above)
Comment (codeofdusk) [reply source/appModules/teamtalk5classic.py]: > When this object is initiated and there is no object, status unavailable is spoken, but creating the object still succeeds. I'd rather have initialisation of the object raise an error, catch that and use the ui.message there instead.

This was from the original add-on, presumably to allow `script_sayToolBarInfo` to be implemented. Can you think of another way to do that now that I've moved the old code over?

File: source/appModules/teamtalk5classic.py (hunk 2 above)
Comment (codeofdusk) [reply source/appModules/teamtalk5classic.py]: > I'm a bit worried about this being a hard refference, introducing a circular refference below when using `self.info = TTToolBarInfo(self)`

I take it this no longer applies now that I've merged the objects?

**Comment by codeofdusk** [comment]
@leonardder I've made an attempt to address some review actions, can you see my comments above?

**Comment by dingpengyu** [comment]
hi @codeofdusk
 appveyor All checks have failed
1 failing check
I can't get the latest test build, can I fix it?
thanks

File: source/appModules/teamtalk5classic.py [hunk 11]
```diff
@@ -0,0 +1,220 @@
...
+	def initOverlayClass(self):
+		n = self.childCount
+		# TeamTalk version-specific toolbar item names by toolbar item count.
+		self.names = []
+		for lst in _msgNames.values():
+			if len(lst) == n:
+				self.names = lst
+				break
+
+	def getToolBar(self):
```
Comment (leonardder) [inline source/appModules/teamtalk5classic.py]: I guess this is now obsolete

File: source/appModules/teamtalk5classic.py [hunk 12]
```diff
@@ -0,0 +1,220 @@
...
+		# First lastChild is a Window object, then the client toolbar object.
+		try:
+			tb = api.getForegroundObject().lastChild.lastChild
+		except AttributeError:
+			tb = None
+		return tb
+
+	def name(self, itemID):
```
Comment (leonardder) [inline source/appModules/teamtalk5classic.py]: This conflicts with the name property

File: source/appModules/teamtalk5classic.py [hunk 13]
```diff
@@ -0,0 +1,220 @@
...
+			return ""
+		return name
+
+
+class TTToolBarItem(IAccessible):
+	def event_valueChange(self):
+		speech.speakObject(self)
+
+	def _get_name(self):
+		return self.parent.info.name(self.IAccessibleChildID)
```
Comment (leonardder) [inline source/appModules/teamtalk5classic.py]: Please fix this as well

File: source/appModules/teamtalk5classic.py [hunk 14]
```diff
@@ -0,0 +1,220 @@
...
+			# Applies to toolbar and its items.
+			if obj.role == controlTypes.ROLE_TOOLBAR:
+				clsList.insert(0, TTToolBar)
+			elif obj.parent and obj.parent.role == controlTypes.ROLE_TOOLBAR:
+				clsList.insert(0, TTToolBarItem)
+
+	@scriptHandler.script(
+		gesture="kb:control+shift+s",
+		# Translators: a gesture description for the TeamTalk voice conferencing client.
+		description=_(
```
Comment (leonardder) [inline source/appModules/teamtalk5classic.py]: Still very long. Please shorten this to a max of 3 sentences, where the first setence just says something like: "Announces relevant information from the Team Talk tool bar" or something like that. Also, to more closely reflect how scripts are documented in core, please change "oon frist press" to something like "pressing once"

File: source/appModules/teamtalk5classic.py [hunk 15]
```diff
@@ -0,0 +1,220 @@
...
+			"Says various on/off states from the toolbar. Depending "
+			"on TeamTalk version, these can include connection status, "
+			"status of voice activation, push-to-talk, and video"
+			"features, whether channel audio is being saved to files,"
+			"and whether the channel message window is showing."
+			"On one press, just says which are checked (in effect)."
+			"On two presses, says all states."
+		)
+	def script_sayToolbarInfo(self, gesture):
+		info = TTToolBarInfo()
```
Comment (leonardder) [inline source/appModules/teamtalk5classic.py]: This won't work now either.

**Review by LeonarddeR — CHANGES_REQUESTED** [review]

**Comment by dingpengyu** [comment]
hello @codeofdusk 
Is the current pr continuing?
thanks

**Comment by dingpengyu** [comment]
hello
The current pr cannot get the test build.
Can you fix PR introduces Flake8 errors?
thanks

**Review by dpy013 — COMMENTED** [review]
Is there any new progress in pr?

File: source/appModules/teamtalk4classic.py [hunk 16]
```diff
@@ -1,23 +1,10 @@
-#appModules/teamtalk4classic.py
-#A part of NonVisual Desktop Access (NVDA)
-#Copyright (C) 2010 NVDA Contributors <http://www.nvda-project.org/>
-#This file is covered by the GNU General Public License.
```
[the remark this answers is not in the document]
Comment (feerrenrut) [reply source/appModules/teamtalk4classic.py:4]: Please see the [wiki for guidelines](https://github.com/nvaccess/nvda/wiki/Copyright-headers)

**Comment by feerrenrut** [comment]
There are a lot of review comments on this PR, it is not clear if they have been resolved.

**Comment by dingpengyu** [comment]
[Accepting bug-fix / maintenance PR's only while addressing backlog](https://github.com/nvaccess/nvda/issues/11006)

**Comment by leonardder** [comment]
@codeofdusk Is this ready for a new review? I don't think so?

**Comment by dingpengyu** [comment]
hello @codeofdusk 
All checks have failed
thanks

**Comment by dingpengyu** [comment]
hi All checks have failed
1 failing check
— AppVeyor build failed
Unable to get latest build to test this pr
thanks

**Comment by dingpengyu** [comment]
cc @codeofdusk 
and
@michaelDCurran 
All checks have failed
— AppVeyor build failed
Lack of access to the latest build for testing.
thanks

**Comment by feerrenrut** [comment]
@codeofdusk There are a number of remaining review actions on this PR. Please take a look and comment / resolve / make changes as appropriate. Then we can move forward with this PR.

**Comment by feerrenrut** [comment]
I'm going to mark this as abandoned, perhaps someone else would like to take over to get this delivered?
